flink-filesystems/flink-gs-fs-hadoop/src/main/java/org/apache/flink/fs/gs/GSFileSystem.java

```diff
@@ -39,17 +45,67 @@
 
     GSFileSystem(GoogleHadoopFileSystem googleHadoopFileSystem, GSFileSystemOptions options) {
         super(Preconditions.checkNotNull(googleHadoopFileSystem));
-        LOGGER.info("Creating GSFileSystem with options {}", options);
-
         this.options = Preconditions.checkNotNull(options);
+        LOGGER.info("Creating GSFileSystem with options {}", options);
     }
 
     @Override
-    public RecoverableWriter createRecoverableWriter() {
-        LOGGER.info("Creating recoverable writer with options {}", options);
+    public RecoverableWriter createRecoverableWriter() throws IOException {
+
+        // follow the same rules as for the Hadoop connector, i.e.
+        // 1) only use service credentials at all if Hadoop
+        // "google.cloud.auth.service.account.enable" is true (default: true)
+        // 2) use GOOGLE_APPLICATION_CREDENTIALS as location of credentials, if supplied
+        // 3) use Hadoop "google.cloud.auth.service.account.json.keyfile" as location of
+        // credentials, if supplied
+        // 4) use no credentials
```

Comment:
Just trying to understand, what happens if a user configs none of this options? I.e, the service account is enabled by default, but no credential is provided.

Reply:
As it stands now, if service account were enabled but no credentials were supplied via either `GOOGLE_APPLICATION_CREDENTIALS` or `google.cloud.auth.service.account.json.keyfile`, it would create a Storage instance with no credential. If you were writing to a publicly writable GCS bucket, this would work, but it would fail if the bucket required credentials.

This is similar to what would happen (as far as I understand) wrt Hadoop config; even if service credentials are enabled (which they are by default), you still have to specify a credential of some kind.

Reply:
Also, it's worth noting that there are more authentication [options](https://github.com/GoogleCloudDataproc/hadoop-connectors/blob/master/gcs/CONFIGURATION.md) supported for Hadoop than just those two:

* Private key and id
* P12 certificate
* Short-lived service-account impersonation

But the only ones that have been documented as supported in Flink so far are the two that are directly mentioned in the new docs, `GOOGLE_APPLICATION_CREDENTIALS` and `google.cloud.auth.service.account.json.keyfile`.

Do you think it's OK just to support those?

Reply:
I think it's fine to only support those mentioned in the documentation, as aligned with the previous behavior.

That leads to another questions: shall we load the entire `core-site/default.xml` and merge them with configurations from flink config? Alternatively we may only look for the configs we supports.

Reply:
Please check me on this, but my thought was that if someone was using the `gcs-connector` before -- i.e. just as a Hadoop-backed FileSystem -- that they would have been able to supply arbitrary config options in the `core-site/default.xml` and expected them to be applied. So I was trying to preserve that behavior.

But, yes, if we support arbitrary config options in `core-site/default.xml` *except* for certain authentication-related options, that does seem a bit counterintuitive.

I suppose one option could be to continue to parse and pass through all the options but to document that the only authentication options that will yield the proper behavior for all FileSystem operations are the two documented, and not the others (P12, etc.).

Reply:
Looking into this a bit more, I think it's probably fine as is.

The module currently consists of two parts leveraging different underlying libaraies: the FileSystem that uses `gcs-connector`, and the RecoverableWriter that uses `google-cloud-storage`. Hadoop configurations (`core-site/default.xml`) can be applied directly on `gcs-connector` but not `google-cloud-storage`.

In that sense, it makes sense to me that RecoverableWriter/`google-cloud-storage` only support selective Hadoop configurations. As a first step, the supported configurations includes only `account.enable` and `account.json.keyfile`. We can add more if new demands emerge later.

WDYT?

Reply:
Yes, I agree.

I'll add a note to the docs (on the other PR) to call this out, and I'll work on some unit tests.